VirtualReality/Widgets/qMRMLVirtualRealityView.cxx

```diff
@@ -549,6 +543,69 @@ void qMRMLVirtualRealityViewPrivate::updateTransformNodeWithHMDPose()
     node->SetAttribute("VirtualReality.HMDActive", "1");
   }
 
+  updateTransformNodeWithPose(node, pose);
+
+  node->EndModify(disabledModify);
+}
+
+//----------------------------------------------------------------------------
+void qMRMLVirtualRealityViewPrivate::updateTransformNodesWithTrackerPoses()
+{
+  Q_Q(qMRMLVirtualRealityView);
+
+  for (uint32_t i = 0; i < this->RenderWindow->GetNumberOfTrackedDevicesForDevice(vtkEventDataDevice::GenericTracker); ++i)
+  {
+    vr::TrackedDeviceIndex_t dev = this->RenderWindow->GetTrackedDeviceIndexForDevice(vtkEventDataDevice::GenericTracker, i);
+    std::stringstream ss;
+    ss << dev;
+
+    bool found(false);
+    vtkMRMLLinearTransformNode* node(nullptr);
+    while (!found)
+    {
+      node = this->MRMLVirtualRealityViewNode->GetNthTrackerTransformNode(i);
+      if (node == nullptr)
+      {
+        break;
+      }
+      std::string id = node->GetAttribute("VirtualReality.VRDeviceID");
```

Comment:
Finding the right transform node should be implemented in the logic or MRML. As things are implemented now, it should be probably in vtkMRMLVirtualRealityViewNode::GetNthTrackerTransformNodeID / Set / Add …
These methods should set and check the VR device ID, too.

Reply:
Do you mean I should implement methods like GetTrackerTransformNodeByVRDeviceID(...)?

Reply:
or instead of using MRML Id for the check, automatically default to the VR device ID? wouldn't that be confusing for other systems that use node id?

Reply:
The most important thing is to have these get/set functions in MRML node (or maybe logic) and not in the widget.

Use node references to keep track of which transform corresponds to which device (similarly as it is done for HMDTransformRole, RightControllerTransformRole, LeftControllerTransformRole). You can use any string for role name, such as VRDeviceID+"TransformRole".

Reply:
I'm having a bit of trouble understanding the optimal architecture.

The widget is currently responsible for populating the MRML node with nodes, and querying for nodes. This is easy because there are a fixed number of nodes (HMD, left controller, right controller).

Now, with N generic trackers, the widget should use Add/Set/GetNthNodeReferenceID with custom roles? What about other modules that may want to access devices by VRID? They'd have to know the same role string construction pattern?

Or should I add accessors to the MRML node Add/Set/GetTransformNodeByVRDeviceID and have those functions contain the role string construction logic?

Reply:
Perhaps a simpler question. Whose job is it to maintain the list of MRML nodes corresponding to generic trackers? Widget, Logic, or MRML?

Reply:
> The widget is currently responsible for populating the MRML node with nodes, and querying for nodes.

This is not the case (if you find anything like this, then it is a bug). Widget is completely optional. Everything should work fine without even instantiating a widget. Widget only offers GUI for the user to modify MRML nodes or launch some functions implemented in the logic.

> Whose job is it to maintain the list of MRML nodes corresponding to generic trackers?

Relationship between nodes can only be stored in MRML nodes (as node references). You can add helper functions in logic that uses information stored in MRML nodes to give different or more convenient API (probably not needed).

> Now, with N generic trackers, the widget should use Add/Set/GetNthNodeReferenceID with custom roles? What about other modules that may want to access devices by VRID?

What is VRID? If it is a string then it can be used as role name as I suggested above (node reference role name: VRDeviceID+"TransformRole"). You can also use a single node reference role to keep reference to multiple nodes (and add an node attribute to each referenced transform node to store the VRDeviceID), but the VRDeviceID+"TransformRole" solution is simpler and more consistent with what we have already. You can get a list of node reference role names from a node and parse it to get the VRDeviceID (we could also store all custom VRDeviceIDs in a custom attribute of vtkMRMLVirtualRealityViewNode, but probably parsing role names is simpler).

Reply:
>This is not the case (if you find anything like this, then it is a bug).

Right now qMRMLVirtualRealityView is the only access into the OpenVR subsystem, thus requiring Widgets for VR to work.

>Relationship between nodes can only be stored in MRML nodes (as node references). You can add helper functions in logic that uses information stored in MRML nodes to give different or more convenient API (probably not needed).

More specifically then, whose job is it to call `SetNthTrackerTransformNode(...)`? 

> What is VRID?
VR ID is the OpenVR subsystem unique ID for a VR device. It is the way to differentiate between different generic trackers.

>node reference role name: VRDeviceID+"TransformRole"

I understand and will implement that, but whose role is it to merge/parse node role names? Should all clients to the MRML class know how to perform the merge/parse or should I have functions inside the MRML node that do this for its clients that hide this knowledge?

Reply:
I think the fundamental problem is quite different. Right now, there is on way to observe VR devices past vtkOpenVR.

One of our students is working to add VTK events to `vtkOpenVRRenderWindowInteractor`. Then, we could have vtkSlicerVirtualRealityLogic observe those events on an interactor and do things properly.

Reply:
> Right now qMRMLVirtualRealityView is the only access into the OpenVR subsystem, thus requiring Widgets for VR to work.

Oh you, mean the view widget. Most of that class should be moved to logic. All we need from the application is to regularly call a the a timer callback in the logic. We should not add logic features there.

> More specifically then, whose job is it to call SetNthTrackerTransformNode(...)?

We can keep it for modules only, without GUI. If you want to allow users to add custom transforms without typing custom scripts (it could be nice), then we could add GUI for it (e.g., a textbox for VR device ID + a node selector to choose transform node + an Add button; a list of current devices; remove button; maybe a pause/resume button).

> I understand and will implement that, but whose role is it to merge/parse node role names? 

vtkMRMLVirtualRealityViewNode

Reply:
>Oh you, mean the view widget. Most of that class should be moved to logic. All we need from the application is to regularly call a the a timer callback in the logic. We should not add logic features there.


Yes, I think that's what was throwing me off.

> >More specifically then, whose job is it to call SetNthTrackerTransformNode(...)?

>We can keep it for modules only, without GUI. If you want to allow users to add custom transforms without typing custom scripts (it could be nice),

To answer my own question, `vtkSlicerVirtualRealityLogic` should observe the interactor (which is set to the MRML node by... the application(GUI or whatever application chooses), and from the interactor events, the logic should keep the MRML node updated with correct node references, yes?

Reply:
We tried to make VTK interactors and widgets work for many years, but it was extremely complicated, slow, and overall too rigid. It has prevented us for many years for having reasonably functional widgets. Now that time is over. 

We completely redesigned how interaction events are handled in Slicer. We now use vtkMRMLViewInteractorStyle to capture all events, convert them to new-type (vtkMRMLInteractionEventData, based on vtkEventData) event and forward them to MRML displayable managers for processing. Displayable managers forward the events to the widgets that they manage for processing.

This design is implemented already for slice and 3D views but not yet for virtual reality view. Since virtual reality controller events are already new-style events, forwarding them to the displayable managers is trivial. What will require some more work is to move the few gestures that are currently implemented in vtkVirtualRealityViewInteractorStyle into a new VTK widget managed by a new displayable manager (very similar to vtkMRMLCameraDisplayableManager and vtkMRMLCameraWidget). Moving out event processing from interactor style is important because otherwise it is extremely difficult to properly route events. For example, with current implementation we cannot highlight a markups or model node when the controller is inside, so the user does not know if pressing a button on the controller will move the object or not. With the new infrastructure, we can just add new event mapping to existing markups widgets to make them react to not just mouse and keyboard but also virtual reality controller events (few lines in vtkSlicerMarkupsWidget constructor).

If all the above is not clear then let's discuss before you start implementing anything.

Reply:
Yes, I think we should chat. This is a larger change than expected and perhaps more than the student should bite off. For now I'll have him produce a local workaround, and we can come up with a game plan for the proper fix.

Reply:
Ok, for this code I'm going to clean it up a bit, but won't be implementing the full displayable manager fix.